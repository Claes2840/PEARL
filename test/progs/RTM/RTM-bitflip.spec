// Inner variables
Q = 'nil
Q1 = 'nil
S1 = 'nil
S2 = 'nil
Q2 = 'nil
RulesRev = 'nil
Rules = 'nil
Rule = 'nil

// Describe the TM
// Flip bits in string (0|1)*
// Example: _101 --> _010
Start = '1
End = '6
pc_max = '9
Transitions = 
 '((1 . (BLANK . (BLANK . 2))) . 
  ((2 . (SLASH . (RIGHT . 3))) . 
  ((3 . (0 .     (1     . 2))) . 
  ((3 . (1 .     (0     . 2))) . 
  ((3 . (BLANK . (BLANK . 4))) . 
  ((4 . (SLASH . (LEFT  . 5))) . 
  ((5 . (0 .     (0     . 4))) . 
  ((5 . (1 .     (1     . 4))) . 
  ((5 . (BLANK . (BLANK . 6))) . 
   nil))))))))

// Tape for full specialization
//S = 'nil
//S_left = 'nil
//S_right = '(1 . (0 . (1 . nil)))
