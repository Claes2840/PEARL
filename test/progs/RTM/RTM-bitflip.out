(S_left S S_right) -> (S_left S S_right)

init_11:
	entry
	assert((S = 'BLANK))
	S ^= 'BLANK
	S ^= 'BLANK
	goto act2_1

act2_1:
	fi (S = 'BLANK)
		from init_11
		else act2_2
	assert(!((S = 'SLASH)))
	if ((S_left = 'nil) && (S = 'BLANK))
		goto right_1b_3
		else right_1p_3

right_1b_3:
	from act2_1
	S ^= 'BLANK
	goto right1_3

right1_3:
	fi (S_left = 'nil)
		from right_1b_3
		else right_1p_3
	if (S_right = 'nil)
		goto right_2b_3
		else right_2p_3

right_2b_3:
	from right1_3
	S ^= 'BLANK
	goto right2_3

right2_3:
	fi ((S_right = 'nil) && (S = 'BLANK))
		from right_2b_3
		else right_2p_3
	if (S = '0)
		goto write_5
		else act2_5

write_5:
	from right2_3
	S ^= '0
	S ^= '1
	assert((S = '1))
	goto act2_2

act2_2:
	fi (S = '0)
		from write_4
		else write_5
	goto act2_1

act2_5:
	from right2_3
	if (S = '1)
		goto write_4
		else act2_4

write_4:
	from act2_5
	S ^= '1
	S ^= '0
	goto act2_2

act2_4:
	from act2_5
	assert((S = 'BLANK))
	S ^= 'BLANK
	S ^= 'BLANK
	goto act2_6

act2_6:
	fi (S = 'BLANK)
		from act2_4
		else act2_7
	assert(!((S = 'SLASH)))
	if ((S_right = 'nil) && (S = 'BLANK))
		goto left_1b_8
		else left_1p_8

left_1b_8:
	from act2_6
	S ^= 'BLANK
	goto left1_8

left1_8:
	fi (S_right = 'nil)
		from left_1b_8
		else left_1p_8
	if (S_left = 'nil)
		goto left_2b_8
		else left_2p_8

left_2b_8:
	from left1_8
	S ^= 'BLANK
	goto left2_8

left2_8:
	fi ((S_left = 'nil) && (S = 'BLANK))
		from left_2b_8
		else left_2p_8
	if (S = '0)
		goto write_10
		else act2_10

write_10:
	from left2_8
	S ^= '0
	S ^= '0
	assert((S = '0))
	goto act2_7

act2_7:
	fi (S = '1)
		from write_9
		else write_10
	goto act2_6

act2_10:
	from left2_8
	if (S = '1)
		goto write_9
		else act2_9

write_9:
	from act2_10
	S ^= '1
	S ^= '1
	goto act2_7

act2_9:
	from act2_10
	assert((S = 'BLANK))
	S ^= 'BLANK
	S ^= 'BLANK
	assert((S = 'BLANK))
	exit

left_2p_8:
	from left1_8
	(S . S_left) <- S_left
	goto left2_8

left_1p_8:
	from act2_6
	S_right <- (S . S_right)
	goto left1_8

right_2p_3:
	from right1_3
	(S . S_right) <- S_right
	goto right2_3

right_1p_3:
	from act2_1
	S_left <- (S . S_left)
	goto right1_3
